File: 3 Longest Substring Without Repeating Characters/3 Longest Substring Without Repeating Characters.swift
class Solution {
  func lengthOfLongestSubstring(_ s: String) -> Int {
    var maxLength: Int = 0
    var startIndex: Int = 0
    var charsMap: [Character:Int] = [:]

    for (i, char) in s.enumerated() {
      if let index = charsMap[char] {
        startIndex = max(startIndex, index + 1)
      }

      charsMap[char] = i
      maxLength = max(maxLength, i - startIndex + 1)
    }

    return maxLength
  }
}

File: 3 Longest Substring Without Repeating Characters/3 Longest Substring Without Repeating Characters.ts
function lengthOfLongestSubstring(s: string): number {
  let maxLength: number = 0;
  let startIndex: number = 0;
  const charsMap: Map<string, number> = new Map();

  for (let i = 0; i < s.length; i++) {
    if (charsMap.has(s[i])) {
      startIndex = Math.max(startIndex, charsMap.get(s[i])! + 1);
    }

    charsMap.set(s[i], i);
    maxLength = Math.max(maxLength, i - startIndex + 1);
  }

  return maxLength;
}
